### PR TITLE
Add retry logic for api calls

### PR DESCRIPTION
**Pull request by muang0**
Retry logic added for consensus & nimbus api calls.

#272

### CODE REVIEW

**Comment by muang0** [comment]
Tested by building & running locally. Happy to look at implementing a more comprehensive testing strategy for these changes if we want.

**Review by ncitron — CHANGES_REQUESTED** [review]
This is looking great!

Can we just do a little deduplicating on some of this code?

I think we can probably have one helper function in `NimbusRpc` with a signature like `async fn get<R: Serialize>(req: &str) -> Result<R>`.

In `CheckpointFallback` we can probably do a similar thing with `async fn get_checkpoints(constructed_url: &str) -> Result<reqwest::Response>`.

**Comment by muang0** [comment]
> This is looking great!
> 
> Can we just do a little deduplicating on some of this code?
> 
> I think we can probably have one helper function in `NimbusRpc` with a signature like `async fn get<R: Serialize>(req: &str) -> Result<R>`.
> 
> In `CheckpointFallback` we can probably do a similar thing with `async fn get_checkpoints(constructed_url: &str) -> Result<reqwest::Response>`.

I'm following you regarding the CheckpointFallback function signature, but I'm struggling to understand the NimbusRpc signature. I'm not sure how we would get the generic funtion to deserialize into the specific type we're looking for ie. `OptimisticUpdateResponse`, `BeaconBlockResponse`, etc. Do we want to handle deserialization of the response outside of the abstracted `get` function similar to in `CheckpointFallback`? Thanks for the help.

**Comment by ncitron** [comment]
I think this is in the right direction, but we should be able to do a bit better.

I think the `.json<SomeOutputType>` part can be modified to `.json<R>` in this example, assuming that `R` is a generic on the `get` function that implements `DeserializeOwned` (judging from the signature of the [json method of reqwest::Response](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.json)).

**Comment by muang0** [comment]
> I think this is in the right direction, but we should be able to do a bit better.
> 
> I think the `.json<SomeOutputType>` part can be modified to `.json<R>` in this example, assuming that `R` is a generic on the `get` function that implements `DeserializeOwned` (judging from the signature of the [json method of reqwest::Response](https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.json)).

Doh, of course! I was getting hung up on how to reference the generic from within the function. Think I've got it in a ready-to-review state now.

File: consensus/src/rpc/nimbus_rpc.rs [hunk 1]
```diff
@@ -28,15 +43,7 @@ impl ConsensusRpc for NimbusRpc {
...
-        let client = reqwest::Client::new();
-        let res = client
-            .get(req)
-            .send()
-            .await
-            .map_err(|e| RpcError::new("bootstrap", e))?
-            .json::<BootstrapResponse>()
-            .await
-            .map_err(|e| RpcError::new("bootstrap", e))?;
+        let res: BootstrapResponse = get(&req).await?;
```
Comment (ncitron) [inline consensus/src/rpc/nimbus_rpc.rs:46]: Looks great. Can we just make sure that the error types here are all `RpcError` like they were originally. It makes it easier when something goes wrong to track from exactly which rpc call it came from.

**Comment by ncitron** [comment]
Oh also can you run a quick `cargo fmt` so that CI passes?

**Pull request merged**